pkg/kv/kvserver/rangefeed/budget.go

```diff
@@ -0,0 +1,249 @@
+	limit int64
+	// Channel to notify that memory was returned to the budget.
+	replenishC chan interface{}
+	// Budget cancellation request
+	stopC chan interface{}
+
+	closed  sync.Once
+	metrics *Metrics
+}
+
+// NewFeedBudget creates a FeedBudget to be used with RangeFeed. If nil account
+// is passed, function will return nil which is safe to use with RangeFeed as
+// it effectively disables memory accounting for that feed.
+func NewFeedBudget(budget *mon.BoundAccount, limit int64, metrics *Metrics) *FeedBudget {
+	if budget == nil {
+		return nil
+	}
+	// If limit is not specified, use large enough value.
+	if limit <= 0 {
+		limit = (1 << 63) - 1
+	}
+	f := &FeedBudget{
+		replenishC: make(chan interface{}, 1),
+		stopC:      make(chan interface{}),
+		metrics:    metrics,
+		limit:      limit,
+	}
+	f.mu.memBudget = budget
+	return f
+}
+
+// Get allocates amount from budget.
+// If budget has less memory than requested it could either wait, if we have any
+// in flight allocations, or it could send a single message through regardless
+// of its size. In latter case, memory is not requested from underlying budget,
+// but we won't be able to do another Get until this message is returned. To
+// track such events, FeedBudget uses a counter metric.
+//
+// timeout argument indicates how long we could wait for the budget to replenish
+// if there is not enough budget available. Budget is replenished by Return
+// calls.
+// timeout value of 0 means wait indefinitely.
+// Whenever we use timeout, we return its wait chan to caller, so it could be
+// used subsequently by the chained wait for other resources thus keeping total
+// timeout in check without adding to it on every op. If we didn't initiate
+// a wait or if timeout was 0 then wait channel will be nil.
+// Returned allocation has its use counter set to 1.
+func (f *FeedBudget) Get(
+	ctx context.Context, amount int64, timeout time.Duration,
+) (*SharedBudgetAllocation, <-chan time.Time, error) {
+	alloc, err := f.tryAlloc(ctx, amount)
+	if err == nil {
+		return getPooledBudgetAllocation(alloc), nil, nil
+	}
+	// If we failed to allocate immediately, then we need to wait on replenish
+	// channel to track Return calls from consumed events.
+	var timeC <-chan time.Time
+	if timeout > 0 {
+		timeC = time.After(timeout)
+	}
+	for {
+		select {
+		case <-f.replenishC:
+			alloc, err = f.tryAlloc(ctx, amount)
+			if err == nil {
+				return getPooledBudgetAllocation(alloc), timeC, nil
+			}
+		case <-timeC:
+			// TODO(oleg): should we try to allocate again in case we are faced
+			// exhausted memory pool and someone returned memory to the pool?
+			f.metrics.RangeFeedBudgetExhausted.Inc(1)
+			return nil, timeC, errors.New("failed to allocate mem budget")
+		case <-ctx.Done():
+			return nil, timeC, nil
+		case <-f.stopC:
+			// We are already stopped, current allocation is already freed so, do
+			// nothing.
+			return nil, timeC, nil
```

Comment:
I understand that perhaps AbstractQuotaPool might be an overkill, but the more I used it, the less I thought it's strange and the more I liked it.  One of the things that AQP does that this code doesn't do, is that could log messages if the allocations are too slow. This functionality proved to be pretty important for debugging -- and something tells me that knowing that allocation is stuck for long time is useful.  @ajwerner might have more input on this; but I just want to bring this up to your attention.

Reply:
Yeah, I was reading it and it seem like a right tool for the job once I manage to get my head around it. I'll create a TODO to see how we could use it.